File: mobile-client/ios/Sources/CoreBilling/Billing.swift
import Foundation

public struct Billing {
  private let uploadReceiptUseCase: UploadReceiptUseCase
  private let verifyReceiptUseCase: VerifyReceiptUseCase
  private let billingReceiptGateway: BillingReceiptGateway

  public init(
    billingReceiptGateway: BillingReceiptGateway = BillingReceiptGateway()
  ) {
    self.billingReceiptGateway = billingReceiptGateway
    self.uploadReceiptUseCase = UploadReceiptUseCase(billingReceiptGateway: billingReceiptGateway)
    self.verifyReceiptUseCase = VerifyReceiptUseCase(billingReceiptGateway: billingReceiptGateway)
  }

  public func uploadReceipt(token: String, product: String) async throws -> Bool {
    try await uploadReceiptUseCase(token: token, product: product)
  }

  public func verify(token: String) async throws -> Bool {
    let result = try await verifyReceiptUseCase(token: token)
    return result.accepted
  }

  public func uploadPayload(token: String, product: String) -> ReceiptUploadPayload {
    ReceiptUploadPayload(token: token, product: product)
  }

  public func defaultVerificationResult(accepted: Bool = false) -> ReceiptVerificationResult {
    ReceiptVerificationResult(accepted: accepted)
  }
}

File: mobile-client/android/core/billing/src/main/java/app/mobiling/core/billing/Billing.kt
package app.mobiling.core.billing

import app.mobiling.client.client.contract.system.billing.ReceiptVerificationResult
import app.mobiling.client.client.data.system.billing.BillingReceiptGateway
import app.mobiling.client.client.usecase.system.billing.UploadReceiptUseCase
import app.mobiling.client.client.usecase.system.billing.VerifyReceiptUseCase

/**
 * Legacy-compatible Android entry point bridged to canonical system/billing slices.
 */
class Billing(
    baseUrl: String = "https://httpbin.org",
) {
    private val billingReceiptGateway: BillingReceiptGateway = BillingReceiptGateway(baseUrl)
    private val uploadReceiptUseCase: UploadReceiptUseCase = UploadReceiptUseCase(billingReceiptGateway)
    private val verifyReceiptUseCase: VerifyReceiptUseCase = VerifyReceiptUseCase(billingReceiptGateway)

    fun uploadReceipt(token: String, product: String): Boolean = uploadReceiptUseCase(token, product)

    fun verifyReceipt(token: String): ReceiptVerificationResult = verifyReceiptUseCase(token)
}
